akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala

```diff
@@ -344,6 +346,53 @@ class ClientServerSpec extends WordSpec with Matchers with BeforeAndAfterAll wit
         }(system2)
         Await.result(b1.unbind(), 1.second)
       }(materializer2)
+
+      "stop stages on failure" in Utils.assertAllStagesStopped {
+        val (_, hostname, port) = TestUtils.temporaryServerHostnameAndPort()
+        val counter = new AtomicLong(0)
+        val stage: GraphStage[FlowShape[HttpRequest, HttpResponse]] = new GraphStage[FlowShape[HttpRequest, HttpResponse]] {
+          val in = Inlet[HttpRequest]("request.in")
+          val out = Outlet[HttpResponse]("response.out")
+
+          override def shape: FlowShape[HttpRequest, HttpResponse] = FlowShape.of(in, out)
+
+          override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+            override def preStart(): Unit = counter.incrementAndGet()
+            override def postStop(): Unit = counter.decrementAndGet()
+            override def onPush(): Unit = push(out, HttpResponse(entity = counter.get().toString))
+            override def onPull(): Unit = pull(in)
+
+            setHandlers(in, out, this)
+          }
+        }
+
+        def performFaultyRequest() = {
+          val socket = new Socket(hostname, port)
+          val os = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream, "UTF8"))
+
+          os.write("YOLO")
+          os.close()
+
+          socket.close()
+
+          Thread.sleep(1000)
```

Comment:
and then use http://www.artima.com/docs-scalatest-2.0.M5/org/scalatest/concurrent/Eventually.html to check the shouldEqual 0